zero-downtime-schema-migrations-postgresql.mdx

```diff
@@ -0,0 +1,119 @@
+
+**Expand & contract**
+
+The expand and contract strategy, also known as [parallel change](https://martinfowler.com/bliki/ParallelChange.html), is a pattern used to migrate your schemas with zero downtime. This pattern is usually seen in decoupled architectures and is proven to work well at scale.
+
+It works by applying changes in a specific series of steps that add in the new schema and data, with the application aware of both schemas for a given period of time. Then, deleting the old schema after the data has been backfilled. This approach is backwards compatible and reversible and a pattern we see used regularly in modern applications.
+
+**Application specific processes**
+
+Every application is unique in one way or another. As applications grow in complexity, so do their schema migrations. You may find yourself writing pages and pages of documentation, but for good reason. Micro-managing schema migrations means that petabytes of customer data are safe and your business continues to run smoothly.
+
+![GitLab migration guide](/images/zdt-schema-gitlab.png)
+
+In larger organizations like this, schema migrations typically are categorized by type and smaller and more frequent changes are preferred. There may also be processes in place for feature flags or canary deployments for multi-tenant environments. Generally, I’d say whatever approach you take – good documentation is always a welcome addition. This [GitLab documentation](https://docs.gitlab.com/ee/development/migration_style_guide.html) is a great example of how organizations tend to handle schema migrations at scale.
+
+**Planned downtime**
+
+Then, there’s always the possibility that downtime is acceptable for your application and business. Maybe your audience is only around during business hours or you have a regular planned maintenance every month. In which case, you can take more of a “big bang” approach. Multi-step process or intimately integrated ORMs simply aren’t necessary to avoid downtime, giving you more time to get creative with your maintenance window page.
+
+## Tools available for schema migrations
+
+When you start seriously considering more destructive changes to your schema, thrid party tooling becomes more of a necessity to keep track of the changes, version them and perform the migrations. Since this problem has been around for awhile, there are a variety of tools. We won't go into details for each one, but some solutions have been around for nearly two decades and others popped up in the last few years.
```

Comment:
```suggestion
When you start seriously considering more destructive changes to your schema, thrid party tooling becomes more of a necessity to keep track of the changes, version them and perform the migrations. Since this problem has been around for a while, there are a variety of tools. We won't go into details for each one, but some solutions have been around for nearly two decades and others have popped up in the last few years.
```